types/wlr_pointer_gestures_v1.c

```diff
@@ -0,0 +1,288 @@
+	.destroy = resource_handle_destroy,
+};
+
+static void get_swipe_gesture(struct wl_client *client,
+		struct wl_resource *gestures_resource,
+		uint32_t id,
+		struct wl_resource *pointer_resource) {
+	struct wlr_pointer_gestures_v1 *gestures =
+		pointer_gestures_from_resource(gestures_resource);
+
+	struct wlr_seat *seat =
+		wlr_seat_client_from_pointer_resource(pointer_resource)->seat;
+
+	struct wl_resource *gesture = wl_resource_create(client,
+		&zwp_pointer_gesture_swipe_v1_interface,
+		wl_resource_get_version(gestures_resource),
+		id);
+	if (gesture == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(gesture, &swipe_impl, seat, resource_remove_from_list);
+	wl_list_insert(&gestures->swipes, wl_resource_get_link(gesture));
+}
+
+void wlr_pointer_send_gesture_pinch_begin(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_seat *seat,
+		uint32_t time_msec,
+		uint32_t fingers) {
+	struct wl_resource *focus = seat->pointer_state.focused_surface->resource;
+	struct wl_resource *gesture;
+	wl_resource_for_each(gesture, &gestures->pinches) {
+		struct wlr_seat *gesture_seat = wlr_seat_from_pointer_resource(gesture);
+		struct wl_client *gesture_client = wl_resource_get_client(gesture);
+		if (gesture_seat != seat || gesture_client != wl_resource_get_client(focus)) {
+			continue;
+		}
+		uint32_t serial = wl_display_next_serial(
+				wl_client_get_display(gesture_client));
+		zwp_pointer_gesture_pinch_v1_send_begin(gesture, serial,
+				time_msec, focus, fingers);
+	}
+}
+
+void wlr_pointer_send_gesture_pinch_update(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_seat *seat,
+		uint32_t time_msec,
+		double dx,
+		double dy,
+		double scale,
+		double rotation) {
+	struct wl_resource *focus = seat->pointer_state.focused_surface->resource;
+	struct wl_resource *gesture;
+	wl_resource_for_each(gesture, &gestures->pinches) {
+		struct wlr_seat *gesture_seat = wlr_seat_from_pointer_resource(gesture);
+		struct wl_client *gesture_client = wl_resource_get_client(gesture);
+		if (gesture_seat != seat || gesture_client != wl_resource_get_client(focus)) {
+			continue;
+		}
+		zwp_pointer_gesture_pinch_v1_send_update(gesture, time_msec,
+				wl_fixed_from_double(dx), wl_fixed_from_double(dy),
+				wl_fixed_from_double(scale),
+				wl_fixed_from_double(rotation));
+	}
+}
+
+void wlr_pointer_send_gesture_pinch_end(
+		struct wlr_pointer_gestures_v1 *gestures,
+		struct wlr_seat *seat,
+		uint32_t time_msec,
+		bool cancelled) {
+	struct wl_resource *focus = seat->pointer_state.focused_surface->resource;
+	struct wl_resource *gesture;
+	wl_resource_for_each(gesture, &gestures->pinches) {
+		struct wlr_seat *gesture_seat = wlr_seat_from_pointer_resource(gesture);
+		struct wl_client *gesture_client = wl_resource_get_client(gesture);
+		if (gesture_seat != seat || gesture_client != wl_resource_get_client(focus)) {
+			continue;
+		}
+		uint32_t serial = wl_display_next_serial(
```

Comment:
Maybe we can use one serial for everybody?

Reply:
Is that fine? I assumed serials must be globally unique…

Reply:
The serial represents the event. It's used as a "timestamp" to prevent races, and also as a token. We'll send it to a single client, so it's fine to generate only one. In fact, it would simplify serial validation, if we need that in the future.